src/fraction/mod.rs

```diff
@@ -2337,15 +2336,54 @@ fraction_from_generic_int!(u8, i8, u16, i16, u32, i32, u64, i64, u128, i128, usi
 macro_rules! generic_fraction_from_float {
     ( $($from:ty),*) => {
         $(
-        impl<T: Clone + Integer + CheckedAdd + CheckedMul + CheckedSub> From<$from> for GenericFraction<T> {
-            fn from (val: $from) -> GenericFraction<T> {
-                if val.is_nan () { return GenericFraction::NaN };
-                if val.is_infinite () { return GenericFraction::Infinity (if val.is_sign_negative () { Sign::Minus } else { Sign::Plus }) };
+        impl<T: Copy + Clone + FromPrimitive + Integer + CheckedAdd + CheckedMul + CheckedSub> From<$from> for GenericFraction<T> {
+            fn from(val: $from) -> GenericFraction<T> {
+                if val.is_nan () { return Self::NaN };
+                if val.is_infinite () { return Self::Infinity (if val.is_sign_negative () { Sign::Minus } else { Sign::Plus }) };
+
+                let sign = if val < 0.0 { Sign::Minus } else { Sign::Plus };
+
+                // Using https://math.stackexchange.com/a/1049723/17452
+                // Find the max precision of this number
+                // Note: all computations happen in i32 until the end.
+                let mut p: i32 = 0;
+                let mut new_val = val;
+                let ten: $from = 10.0;
+                loop {
+                    if (new_val.floor() - new_val).abs() < <$from>::EPSILON {
+                        // Yay, we've found the precision of this number
+                        break;
+                    }
+                    // Multiply by the precision
+                    // Note: we multiply by powers of ten to avoid this kind of round error with f32s:
+                    // https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b760579f103b7192c20413ebbe167b90
+                    p += 1;
+                    new_val = val * ten.powi(p);
```

Comment:
`new_val` becomes `inf` here on capacity overflow. We'll have to check for it and fall back to the `string` conversion if that happens.